include/deal.II/base/symbolic_function.h

```diff
@@ -0,0 +1,315 @@
+
+#ifndef dealii_symbolic_function_h
+#define dealii_symbolic_function_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/tensor.h>
+
+#include <deal.II/differentiation/sd.h>
+
+#include <functional>
+#include <iostream>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+#ifndef DOXYGEN
+template <typename number>
+class Vector;
+namespace Functions
+{
+  template <int dim, typename RangeNumberType>
+  class SymbolicFunction;
+}
+#endif
+
+namespace Functions
+{
+#ifdef DEAL_II_WITH_SYMENGINE
+  /**
+   * A Function class that leaverages symbolic differentiation to compute
+   * gradients, laplacians, hessians, and time derivatives.
+   *
+   * This class can be used to define functions using methods provided by the
+   * Differentiation::SD namespace. In particular, one can define a symbolic
+   * evaluation point (the argument of the function), as well as a symbolic
+   * expression.
+   *
+   * The symbolic gradients and the symbolic Hessians are computed at
+   * construction time, and when a substitution in the symbolic functions is
+   * requested by the user using the method apply_substitution_map().
+   *
+   * Whenever one of the evaluation methods is called, a substitution is
+   * attempted with the symbolic argument replaced by the evaluation point.
+   * The user has to make sure that at evaluation time argument substitution
+   * provides a fully evaluated expression (i.e., no other symbols are contained
+   * in the function expression, except numerical values), or an exception will
+   * be thrown. Additional symbols can be partially evaluated or substituted
+   * by calling apply_substitution_map().
+   *
+   * The simplest use case of this class is given in the following example:
+   * @code
+   * SymbolicFunction<2> fun("x^2+y; t*x*y");
+   * fun.set_time(3.0);
+   * Point<2> p(1.0, 2.0);
+   *
+   * auto a = fun.value(p, 0); // a = 3.0
+   * auto b = fun.value(p, 1); // b = 6.0
+   *
+   * auto df_dt = fun.time_derivative();
+   *
+   * auto c = df_dt.value(p, 0); // c = 0.0
+   * auto d = df_dt.value(p, 1); // d = 2.0
+   * @endcode
+   * where a Function with two components is defined using a string containing
+   * their expressions separated by semicolons.
+   *
+   * A more involved example, that explicitly uses
+   * Differentiation::SD::Expression() objects, is given by
+   * @code
+   * // Create a Tensor<1,2,Differentiation::SD::Expression>
+   * // with symbols "x" and "y", and the symbol "t"
+   * const auto x = SymbolicFunction<2>::default_argument();
+   * const auto t = Differentiation::SD::make_symbol("t");
+   *
+   * // Use directly x[0] (the symbol "x"), x[1] (the symbol "y"), and t
+   * // (the symbol "t").
+   * auto f = std::sin(x[0])*std::cos(x[1])*std::sin(t);
```

Comment:
It's really just a `double`. Let's make that clear to prevent people from trying to think too hard:
```suggestion
   * const double f = std::sin(x[0])*std::cos(x[1])*std::sin(t);
```

Reply:
Or did you mean to enclose this in `"..."`?

Reply:
It is *not* just a double. `x[0]` is of type `SD::Expression`, and so is `x[1]` and `t`. The result is a symbolic expression.

Reply:
Or, to be more precise, the result is a `Differentiation::SD::Expression` object.

Reply:
Oh, I see. Can you make that clear in the comment above (unless it already is and I just didn't read carefully enough).

Reply:
Done.